File: ruby/json/json_simple_analyzer.rb
# -*- coding: utf-8 -*-

require 'json'
require 'awesome_print'

class Analyzer
  def initialize(args)
    @filename = args.shift || "json.txt"
    @hash = Hash.new
  end

  def start
    open(@filename) do |file|
      file.each_line do |line|
        key, tag, json = line.force_encoding("utf-8").strip.split("\t")
        JSON.parse(json).each {|word, count|
          @hash.has_key?(word) ? @hash[word] += count : @hash[word] = count
        }
      end
    end

    i = 0
    @hash.sort_by{|k, v| -v}.each {|k, v|
      i += 1
      output(i, k, v)
    }
  end

  private

  def output(key, tag, value)
    puts "#{key}\t#{tag}\t#{value}"
  end
end

if __FILE__ == $0
  analyzer = Analyzer.new(ARGV)
  analyzer.start
end

File: python/json/json_simple_analyzer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os
import json
from collections import OrderedDict

class Analyzer:

    def __init__(self, args):
        self.filename = args[1]
        self.dic = OrderedDict()

    def start(self):
        file = open(self.filename, 'r')
        for line in file:
            key, tag, value = line.rstrip().split("\t")
            json_obj = json.loads(value)
            for word, count in json_obj.items():
                if word in self.dic:
                    self.dic[word] += count
                else:
                    self.dic[word] = count

        i = 0
        for k, v in sorted(self.dic.items(), key=lambda x: int(x[1]), reverse=True):
            i += 1
            self.__output(i, k, v)
        file.close
        return self.dic

    def __output(self, key, tag, value):
        print(key, tag, value, sep="\t")

if __name__ == '__main__':
    argsmin = 1
    version = (3, 0)
    if sys.version_info > (version):
        if len(sys.argv) > argsmin:
            analyzer = Analyzer(sys.argv)
            analyzer.start()
        else:
            print("This program needs at least %(argsmin)s arguments" %
                  locals())
    else:
        print("This program requires python > %(version)s" % locals())
